### PR TITLE
[test] Add more complicated tests for building and destroying SNodeTrees

### PR DESCRIPTION
**Pull request by ysh329**
## Status: waiting for review

Related issue = #2724 

1. Add more compliated test for destroying and building of field builder:
    1. test field with shape;
    2. test fields builder dense;
    3. test_fields builder pointer;
    4. test fields builder destroy.
2. Re sort out and rename existing tests.

Please give some advice etc.

### CODE REVIEW

File: tests/python/test_fields_builder.py [hunk 1]
```diff
@@ -3,174 +3,259 @@
...
+
+    def create_fb():
+        fb = ti.FieldsBuilder()
+        x = ti.field(ti.f32)
+        fb.dense(ti.ij, shape_size).place(x)
+        fb.finalize()
+
+    # kMaxNumSnodeTreesLlvm=32 in taichi/inc/constants.h
+    # TODO(ysh329): kMaxNumSnodeTreesLlvm=512 not 32 in `taichi/inc/constants.h`.
+    #               Can this value `kMaxNumSnodeTreesLlvm` load from `.h` file?
```
Comment (ysh329) [inline tests/python/test_fields_builder.py:255]: kMaxNumSnodeTreesLlvm=512 not 32 in `taichi/inc/constants.h`.
               Can this value `kMaxNumSnodeTreesLlvm` load from `.h` file?

File: tests/python/test_fields_builder.py [hunk 2]
```diff
@@ -3,174 +3,259 @@
...
+
+                        fb0.dense(ti.i, test_1d_size).place(field0)
+                        fb1.pointer(ti.i, test_1d_size).place(field1)
+
+                        fb0_snode_tree = fb0.finalize()
+                        fb1_snode_tree = fb1.finalize()
+
+        # destroy
+        fb0_snode_tree.destroy()
+        fb1_snode_tree.destroy()
```
Comment (ysh329) [inline tests/python/test_fields_builder.py:220]: Is here a proper destory place?

File: tests/python/test_fields_builder.py [hunk 3]
```diff
@@ -3,174 +3,259 @@
...
 
-    def B(i):
-        n = i * 10**3
-        fb = ti.FieldsBuilder()
-        a = ti.field(ti.f64)
-        fb.dense(ti.i, n).place(a)
-        c = fb.finalize()
+        with pytest.raises(InvalidOperationError) as e:
+            id_1 = c.id()
+            c.destroy()
```
Comment (ysh329) [inline tests/python/test_fields_builder.py:235]: It seems it can't be triggered?

**Comment by ysh329** [comment]
I found a python format error as blew from link(https://github.com/taichi-dev/taichi/runs/4135911412?check_suite_focus=true):
image

File: tests/python/test_fields_builder.py [hunk 4]
```diff
@@ -3,174 +3,276 @@
...
 
 @ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan, ti.metal])
 def test_fields_with_shape():
-    n = 5
-    x = ti.field(ti.f32, [n])
 
+    # Initialize common variables for tests below
+    shape_size_1d = 5
+    x = ti.field(ti.f32, [shape_size_1d])
+    y = ti.field(ti.f32, [shape_size_1d])
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:17]:
```suggestion
    x = ti.field(ti.f32, shape=shape_size_1d)
    y = ti.field(ti.f32, shape=shape_size_1d)
```
This is more common practice in Taichi.

File: tests/python/test_fields_builder.py (hunk 1 above)
Comment (strongoier) [reply tests/python/test_fields_builder.py:272]: This test has been removed due to long execution time :-)

File: tests/python/test_fields_builder.py (hunk 2 above)
Comment (strongoier) [reply tests/python/test_fields_builder.py:235]: It is OK to destroy here. However, you cannot call `.dense()` or `.pointer()` after calling `.finalize()`. This will trigger `InvalidOperationError`.

File: tests/python/test_fields_builder.py (hunk 3 above)
Comment (strongoier) [reply tests/python/test_fields_builder.py:250]: `id` is a property instead of a method.

**Review by strongoier — COMMENTED** [review]
Thanks for contributing! A little suggestion: try to run locally first :-)

File: tests/python/test_fields_builder.py [hunk 5]
```diff
@@ -3,174 +3,276 @@
...
+
+    # [shape] 1. Test and validation for assign once kernel function
     @ti.kernel
-    def func():
-        for i in range(n):
+    def assign_once_kernel_func():
```
Comment (qiao-bo) [inline tests/python/test_fields_builder.py:21]: maybe just `assign_field`? `*_kernel_func` is a bit confusing.

File: tests/python/test_fields_builder.py [hunk 6]
```diff
@@ -3,174 +3,276 @@
...
 def test_fields_with_shape():
-    n = 5
-    x = ti.field(ti.f32, [n])
 
+    # Initialize common variables for tests below
+    shape_size_1d = 5
+    x = ti.field(ti.f32, [shape_size_1d])
+    y = ti.field(ti.f32, [shape_size_1d])
+
+    # [shape] 1. Test and validation for assign once kernel function
```
Comment (qiao-bo) [inline tests/python/test_fields_builder.py:19]:
```suggestion
    # [shape] 1. Test and validation for assigning fields
```

File: tests/python/test_fields_builder.py (hunk 5 above)
Comment (qiao-bo) [reply tests/python/test_fields_builder.py:21]: same for all occurrences below

File: tests/python/test_fields_builder.py [hunk 7]
```diff
@@ -3,174 +3,276 @@
...
+'''
+Test fields with builder dense.
+'''
+
+
 @ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan, ti.metal])
 def test_fields_builder_dense():
-    n = 5
 
+    # Initialize common variables for tests below
```
Comment (qiao-bo) [inline tests/python/test_fields_builder.py:56]: I think the comment in this line is unnecessary

Comment (qiao-bo) [inline tests/python/test_fields_builder.py:56]: I think the comment in this line is unnecessary, same for other places

File: tests/python/test_fields_builder.py [hunk 8]
```diff
@@ -3,174 +3,276 @@
...
+
 @ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan])
 def test_fields_builder_destroy():
-    def A(i):
-        n = i * 10**3
+
+    # [destroy] Args of main body
+    test_sizes = [1]  # [1, 2, 3]
+    size_extend_factor = 1  # 10 ** 3
+    # note: currently only consider preicison all paltform supported,
```
Comment (qiao-bo) [inline tests/python/test_fields_builder.py:183]:
```suggestion
    # note: currently only consider precision all platform supported,
```

File: tests/python/test_fields_builder.py [hunk 9]
```diff
@@ -3,174 +3,276 @@
...
+    # more detailed here: https://docs.taichi.graphics/lang/articles/basic/type#supported-primitive-types
+    field_types = [ti.f32, ti.i32]
+    field_sizes = [1]  # [1, 5, 10]
+
+    # [destroy] 1. test for single destroy multiple fields
```
Comment (qiao-bo) [inline tests/python/test_fields_builder.py:188]: the function name below basically convey the same information ;). this is why we write function name like this. I don't think we need this comment as well, same as all comments above.

**Review by qiao-bo — COMMENTED** [review]
Thanks for contributing!

File: tests/python/test_fields_builder.py (hunk 1 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:272]: OKay

File: tests/python/test_fields_builder.py (hunk 3 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:250]: Ok, I'll remove it

File: tests/python/test_fields_builder.py (hunk 4 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:17]: OK

File: tests/python/test_fields_builder.py (hunk 6 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:19]: OK

File: tests/python/test_fields_builder.py (hunk 5 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:21]: fixed

File: tests/python/test_fields_builder.py (hunk 7 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:56]: Fixed

Comment (ysh329) [reply tests/python/test_fields_builder.py:56]: Fixed

File: tests/python/test_fields_builder.py (hunk 8 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:183]: Fixed

File: tests/python/test_fields_builder.py (hunk 9 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:188]: Fixed

File: tests/python/test_fields_builder.py (hunk 2 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:235]: Ok

**Comment by qiao-bo** [comment]
Hi, can you please fix the errors? thanks

**Comment by ysh329** [comment]
> Hi, can you please fix the errors? thanks

Of course, however, i found i can't reproduce these errors. My terminal shows like below:

```shell
~/code/taichi$ python tests/python/test_fields_builder.py
[Taichi] version 0.8.3, llvm 10.0.0, commit 27bb1dc3, linux, python 3.9.5
[W 11/19/21 09:22:09.949 817] [cuda_driver.cpp:CUDADriver@38] CUDA driver not found.
```

Any idea?

**Comment by qiao-bo** [comment]
> > Hi, can you please fix the errors? thanks
> 
> Of course, however, i found i can't reproduce these errors. My terminal shows like below:
> 
> ```shell
> ~/code/taichi$ python tests/python/test_fields_builder.py
> [Taichi] version 0.8.3, llvm 10.0.0, commit 27bb1dc3, linux, python 3.9.5
> [W 11/19/21 09:22:09.949 817] [cuda_driver.cpp:CUDADriver@38] CUDA driver not found.
> ```
> 
> No other logs about any error. Any idea?

is your CUDA installed correctly? and also did you enable CUDA when building Taichi? you can attach the output of `ti diagnose` and `nvidia-smi` here

**Comment by ysh329** [comment]
> > > Hi, can you please fix the errors? thanks
> > 
> > 
> > Of course, however, i found i can't reproduce these errors. My terminal shows like below:
> > ```shell
> > ~/code/taichi$ python tests/python/test_fields_builder.py
> > [Taichi] version 0.8.3, llvm 10.0.0, commit 27bb1dc3, linux, python 3.9.5
> > [W 11/19/21 09:22:09.949 817] [cuda_driver.cpp:CUDADriver@38] CUDA driver not found.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > No other logs about any error. Any idea?
> 
> is your CUDA installed correctly? and also did you enable CUDA when building Taichi? you can attach the output of `ti diagnose` and `nvidia-smi` here

Is nvidia card necessary? 😢

**Comment by strongoier** [comment]
You can test CPU only. Please merge with current master and use `python3 tests/run_tests.py test_fields_builder` to run your test.

**Comment by ysh329** [comment]
> You can test CPU only. Please merge with current master and use `python3 tests/run_tests.py test_fields_builder` to run your test.

OKay

**Comment by ysh329** [comment]
Hi @strongoier it seems those workflow is waiting for a night.

**Comment by ysh329** [comment]
太难了

File: tests/python/test_fields_builder.py [hunk 10]
```diff
@@ -3,174 +3,226 @@
...
+                                             field_types, field_sizes):
         fb = ti.FieldsBuilder()
-        a = ti.field(ti.f64)
-        fb.dense(ti.i, n).place(a)
-        c = fb.finalize()
-        c.destroy()
-
-    def B(i):
-        n = i * 10**3
+        for tsize_idx in range(len(test_sizes)):
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:172]: What about following https://docs.taichi.graphics/lang/articles/contribution/write_test#parametrize-test-inputs for parametrized inputs?

File: tests/python/test_fields_builder.py [hunk 11]
```diff
@@ -3,174 +3,226 @@
...
+Test fields with builder destory.
+'''
+
+
 @ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan])
 def test_fields_builder_destroy():
-    def A(i):
-        n = i * 10**3
+    test_sizes = [1]  # [1, 2, 3]
+    size_extend_factor = 1  # 10 ** 3
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:163]: Can we remove `size_extend_factor` and directly have them in `sizes`?

File: tests/python/test_fields_builder.py [hunk 12]
```diff
@@ -3,174 +3,226 @@
...
+'''
+Test fields with builder destory.
+'''
+
+
 @ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan])
 def test_fields_builder_destroy():
-    def A(i):
-        n = i * 10**3
+    test_sizes = [1]  # [1, 2, 3]
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:162]: We shouldn't have these comments in final version :-)

File: tests/python/test_fields_builder.py [hunk 13]
```diff
@@ -3,174 +3,226 @@
...
+Test fields with shape.
+'''
+
 
 @ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan, ti.metal])
 def test_fields_with_shape():
-    n = 5
-    x = ti.field(ti.f32, [n])
+    shape_size_1d = 5
+    x = ti.field(ti.f32, shape=shape_size_1d)
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:14]: Can we simply use `shape` instead of `shape_size_1d`?

**Review by strongoier — COMMENTED** [review]
Thanks for your efforts!

File: tests/python/test_fields_builder.py (hunk 13 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:14]: fixed

File: tests/python/test_fields_builder.py (hunk 11 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:163]: fixed

File: tests/python/test_fields_builder.py (hunk 12 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:162]: fixed

File: tests/python/test_fields_builder.py (hunk 10 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:171]: 1. 装饰器修饰的函数支持两层嵌套吗？因为我看是在内层的一个test函数需要加装饰器；
```python
@ti.test()
def test_log10(x, y, z):
    r = ti.field(ti.f32, ())

    @ti.test()
    def foo():
        pass
```
2. 多个参数作为装饰器可以这么写吗？
```python
@pytest.mark.parametrize('x', [1, 10, 100])
@pytest.mark.parametrize('y', [1, 10, 100])
@pytest.mark.parametrize('z', [1, 10, 100])
@ti.test()
def test_log10(x, y, z):
    r = ti.field(ti.f32, ())

    @ti.kernel
    def foo():
        r[None] = ti.log10(r[None])

    r[None] = x + y + z
    foo()
    assert r[None] == math.log10(x + y + z)
```

Comment (qiao-bo) [reply tests/python/test_fields_builder.py:171]: Hi, yes, you can write it as shown in 2. Check some of our [existing tests](https://github.com/taichi-dev/taichi/blob/7d082fc9def0c1a3c6472363e84aeffa5fe5ab8a/tests/python/test_ndarray.py#L29) for how to use parameterized inputs.

Nevertheless, i strongly suggest to have a local environment for building and testing.

Comment (ysh329) [reply tests/python/test_fields_builder.py:171]: Yes, but I have no NVDIA or AMD GPU. 😢

Comment (ysh329) [reply tests/python/test_fields_builder.py]: Fixed and trying in CI.

**Comment by ysh329** [comment]
Hi @strongoier, it seems CI is stucked.

**Comment by strongoier** [comment]
Would you merge with master and try again?

**Comment by ysh329** [comment]
> Would you merge with master and try again?
No problem, let me try again!

**Comment by ysh329** [comment]
Hi @strongoier @k-ye @qiao-bo plz review again~

File: tests/python/test_fields_builder.py [hunk 14]
```diff
@@ -3,174 +3,205 @@
 
 import taichi as ti
 
+'''
+Test fields with shape.
+'''
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:8]: These comments seem redundant.

File: tests/python/test_fields_builder.py [hunk 15]
```diff
@@ -3,174 +3,205 @@
...
 
-@ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan, ti.metal])
+
+@ti.test(arch=ti.get_host_arch_list())
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:11]: The original archs (`ti.cpu, ti.cuda, ti.vulkan, ti.metal`) for testing should not be removed.

File: tests/python/test_fields_builder.py [hunk 16]
```diff
@@ -3,174 +3,205 @@
...
 def test_fields_builder_destroy():
-    def A(i):
-        n = i * 10**3
+    # note: currently only consider precison that all platform supported,
+    # more detailed here: https://docs.taichi.graphics/lang/articles/basic/type#supported-primitive-types
+    @pytest.mark.parametrize('test_1d_size', [1, 10, 100])
+    @pytest.mark.parametrize('field_type', [ti.f32, ti.i32])
+    @ti.test(arch=ti.get_host_arch_list())
+    def test_for_single_destroy_multi_fields():
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:167]: I suggest putting these tests to the top level.

**Review by strongoier — CHANGES_REQUESTED** [review]

File: tests/python/test_fields_builder.py (hunk 14 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:8]: fixed

File: tests/python/test_fields_builder.py (hunk 15 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:11]: I refered other tests and changed from `ti.cpu, ti.cuda, ti.vulkan, ti.metal` to `arch=ti.get_host_arch_list()`.

Comment (ysh329) [reply tests/python/test_fields_builder.py:11]:
```python
def get_host_arch_list():
    return [_ti_core.host_arch()]
```

However I not found the python impl. of the `host_arch()`, it seems a C++ method?

File: tests/python/test_fields_builder.py (hunk 16 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:162]: fixed

File: tests/python/test_fields_builder.py (hunk 15 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:10]: I changed back from `arch=ti.get_host_arch_list()` to `ti.cpu, ti.cuda, ti.vulkan, ti.metal`

File: tests/python/test_fields_builder.py (hunk 14 above)
Comment (strongoier) [reply tests/python/test_fields_builder.py:8]: Did you push the correct version? It seems that these lines are still there :-)

File: tests/python/test_fields_builder.py [hunk 17]
```diff
@@ -1,176 +1,197 @@
...
 
-@ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan])
+'''
+Test fields with builder destory.
+'''
+
+@pytest.mark.parametrize('test_1d_size', [1, 10, 100])
+@pytest.mark.parametrize('field_type', [ti.f32, ti.i32])
+@ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan, ti.metal])
 def test_fields_builder_destroy():
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:159]:
```suggestion
def test_fields_builder_destroy(test_1d_size, field_type):
```

**Review by strongoier — CHANGES_REQUESTED** [review]

File: tests/python/test_fields_builder.py (hunk 17 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py]: Fixed. 😢

File: tests/python/test_fields_builder.py (hunk 14 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py]: Fixed and I removed all redundant notes.

File: tests/python/test_fields_builder.py [hunk 18]
```diff
@@ -6,171 +6,177 @@
...
 
-@ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan])
-def test_fields_builder_destroy():
-    def A(i):
-        n = i * 10**3
+@pytest.mark.parametrize('test_1d_size', [1, 10, 100])
+@pytest.mark.parametrize('field_type', [ti.f32, ti.i32])
+@ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan, ti.metal])
+def test_fields_builder_destroy(test_1d_size, field_type):
+    # note: currently only consider precison that all platform supported,
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:145]:
```suggestion
    # We currently only consider data types that all platforms support.
```

File: tests/python/test_fields_builder.py [hunk 19]
```diff
@@ -6,171 +6,177 @@
...
 
-    func1()
-    for i in range(n):
+    assign_field_single()
+    for i in range(shape):
         assert x[i] == i * 3
 
 
-@ti.test(arch=[ti.cpu, ti.cuda, ti.metal])
+@ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan, ti.metal])
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:83]:
```suggestion
@ti.test(arch=[ti.cpu, ti.cuda, ti.metal])
```
You can see CI fails. This is because the Vulkan backend doesn't support `pointer` yet.

File: tests/python/test_fields_builder.py [hunk 20]
```diff
@@ -6,171 +6,177 @@
...
-@ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan])
-def test_fields_builder_destroy():
-    def A(i):
-        n = i * 10**3
+@pytest.mark.parametrize('test_1d_size', [1, 10, 100])
+@pytest.mark.parametrize('field_type', [ti.f32, ti.i32])
+@ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan, ti.metal])
+def test_fields_builder_destroy(test_1d_size, field_type):
+    # note: currently only consider precison that all platform supported,
+    # more detailed here: https://docs.taichi.graphics/lang/articles/basic/type#supported-primitive-types
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:146]:
```suggestion
    # See https://docs.taichi.graphics/lang/articles/basic/type#supported-primitive-types for more details.
```

BTW could you move these comments before the test?

**Review by strongoier — CHANGES_REQUESTED** [review]

File: tests/python/test_fields_builder.py (hunk 18 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py]: Fixed.

File: tests/python/test_fields_builder.py (hunk 19 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py]: Yeah. Fixed.

File: tests/python/test_fields_builder.py (hunk 20 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py]: Fixed.

File: tests/python/test_fields_builder.py [hunk 21]
```diff
@@ -6,171 +6,176 @@
...
-    fb2.pointer(ti.i, n).place(z)
+    fb2.pointer(ti.i, shape).place(z)
     fb2.finalize()
 
-    # test range-for
     @ti.kernel
-    def func2():
-        for i in range(n):
+    def assign_field_multiple0():
+        # test range-for
```
Comment (qiao-bo) [inline tests/python/test_fields_builder.py:109]: maybe move this comment back to the original position? func+assert combined is the test.

File: tests/python/test_fields_builder.py [hunk 22]
```diff
@@ -6,171 +6,176 @@
...
     z = ti.field(ti.f32)
-    fb2.pointer(ti.i, n).place(z)
+    fb2.pointer(ti.i, shape).place(z)
     fb2.finalize()
 
-    # test range-for
     @ti.kernel
-    def func2():
-        for i in range(n):
+    def assign_field_multiple0():
```
Comment (qiao-bo) [inline tests/python/test_fields_builder.py:108]: if you want to change the name. Here better to be `assign_field_multiple_range_for` instead of `...0`. Likewise for the name below, better be struct_for

File: tests/python/test_fields_builder.py [hunk 23]
```diff
@@ -6,171 +6,176 @@
...
-        c.destroy()
-        ci.destroy()
+        fb0_snode_tree = fb0.finalize()
+        fb1_snode_tree = fb1.finalize()
+
+        # destroy
+        fb0_snode_tree.destroy()
+        fb1_snode_tree.destroy()
+
+    def test_for_raise_twice_destroy():
```
Comment (qiao-bo) [inline tests/python/test_fields_builder.py:172]: test_for_raise_destroy_twice ?

File: tests/python/test_fields_builder.py (hunk 22 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:108]: A good advice! Fixed.

File: tests/python/test_fields_builder.py (hunk 23 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:173]: Fixed.

**Comment by ysh329** [comment]
XDM @strongoier @qiao-bo 😭 , it seems CI is stucked.

File: tests/python/test_fields_builder.py [hunk 24]
```diff
@@ -6,171 +6,175 @@
...
             x[i] = i
 
-    func()
-
-    for i in range(n):
+    assign_field_single()
+    for i in range(shape):
         assert x[i] == i
 
-    y = ti.field(ti.f32, [n])
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:22]: This line shouldn't be moved to the top - it implicitly uses the field builder.

File: tests/python/test_fields_builder.py [hunk 25]
```diff
@@ -6,171 +6,175 @@
...
+    assign_field_single()
+    for i in range(shape):
         assert x[i] == i * 3
 
 
-@ti.test(arch=[ti.cpu, ti.cuda, ti.vulkan])
-def test_fields_builder_destroy():
-    def A(i):
-        n = i * 10**3
+# We currently only consider data types that all platform support,
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:139]:
```suggestion
# We currently only consider data types that all platforms support.
```

File: tests/python/test_fields_builder.py [hunk 26]
```diff
@@ -6,171 +6,175 @@
...
-        fbi = ti.FieldsBuilder()
-        ai = ti.field(ti.f64)
-        fbi.dense(ti.i, n).place(ai)
-        ci = fbi.finalize()
+        fb0_snode_tree = fb0.finalize()
+        fb1_snode_tree = fb1.finalize()
 
-        c.destroy()
-        ci.destroy()
+        # destroy
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:167]:
```suggestion
```

File: tests/python/test_fields_builder.py [hunk 27]
```diff
@@ -6,171 +6,175 @@
...
+        fb = ti.FieldsBuilder()
+        a = ti.field(ti.f32)
+        fb.dense(ti.i, test_1d_size).place(a)
+        c = fb.finalize()
 
-    for i in range(5):
-        A(5)
-    B(2)
-    A(4)
+        with pytest.raises(InvalidOperationError) as e:
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:177]:
```suggestion
        with pytest.raises(InvalidOperationError):
```

File: tests/python/test_fields_builder.py [hunk 28]
```diff
@@ -6,171 +6,175 @@
...
+            # Triggered if destroy twice
```
Comment (strongoier) [inline tests/python/test_fields_builder.py:178]:
```suggestion
```

**Review by strongoier — CHANGES_REQUESTED** [review]

File: tests/python/test_fields_builder.py (hunk 25 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py]: Fixed.

File: tests/python/test_fields_builder.py (hunk 27 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py]: Fixed.

File: tests/python/test_fields_builder.py (hunk 28 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py]: Fixed.

File: tests/python/test_fields_builder.py (hunk 26 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py]: Fixed.

File: tests/python/test_fields_builder.py (hunk 24 above)
Comment (ysh329) [reply tests/python/test_fields_builder.py:22]: However, these two variables `x` and `y` are used by multi-functions.

Comment (strongoier) [reply tests/python/test_fields_builder.py:22]: `assign_field_single` doesn't use `y` so it's fine to leave `y` here right?

Comment (ysh329) [reply tests/python/test_fields_builder.py:22]: Oh, I know your means.

Comment (ysh329) [reply tests/python/test_fields_builder.py:22]: Fixed.

**Review by strongoier — APPROVED** [review]
LGTM! Thanks for your efforts and patience for finally finishing this! :-)

**Review by qiao-bo — APPROVED** [review]
LGTM! and thanks for the work!
